### PR TITLE
graceful training cleanup after KeyboardInterrupt

### LINKED ISSUE #795: support graceful training cleanup after KeyboardInterrupt
## 🚀 Feature
In the training loop, we should catch `KeyboardInterrupt` exceptions and enter a graceful termination (ie. run `on_training_end`, print profiler report, etc). 

### Motivation

Sometimes I'll train a model and just let it run until something interesting happens or my loss has converged. Exiting via "Ctrl + C" can be easier than waiting for "EarlyStopping" to trigger in some cases. However, I may still want certain things to occur (eg. printing out the profiler summary) after stopping training. 

Salesforce's AWD-LSTM has an example of this feature implemented [here](https://github.com/salesforce/awd-lstm-lm/blob/master/main.py#L287). 

### Pitch

Figure out what actions we want to run after a KeyboardInterrupt. Then adjust the training loop to catch this exception and ensure the aforementioned actions still run afterwards.

### PR DESCRIPTION
**Pull request by jeremyjordan**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #795 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by jeremyjordan** [comment]
We should be able to close #657 as well

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -903,9 +903,6 @@ def run_pretrain_routine(self, model):
         # CORE TRAINING LOOP
         self.train()
 
-        # summarize profile results
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: Does it mean we do not want to show the stat always?
Shall the `run_training_teardown()` be called also here?

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -285,90 +286,87 @@ def train(self):
...
-
-        with self.profiler.profile('on_train_end'):
-            model.on_train_end()
-
-        if self.logger is not None:
-            self.logger.finalize("success")
+        try:
+            # run all epochs
+            for epoch in range(self.current_epoch, self.max_epochs):
+                # set seed for distributed sampler (enables shuffling for each epoch)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]: it is quite a heavy loop, I would personally cut out the loop body into a function... it would make it easier to read/navigate what is loop level and what try...

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -285,90 +286,87 @@ def train(self):
...
-                    return
-
-        self.main_progress_bar.close()
-
-        with self.profiler.profile('on_train_end'):
-            model.on_train_end()
-
-        if self.logger is not None:
-            self.logger.finalize("success")
+        try:
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]: btw, is there any other change then encapsulating the train loop into Try/Except?

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -285,90 +286,87 @@ def train(self):
...
+                    # stop training
+                    stop = should_stop and met_min_epochs
+                    if stop:
+                        self.run_training_teardown()
+                        return
+
+            self.run_training_teardown()
+        except KeyboardInterrupt:
+            log.info('Detected KeyboardInterrupt, attempting graceful shutdown...')
+            self.run_training_teardown()
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]: :+1:

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -574,6 +572,20 @@ def optimizer_closure():
 
         return 0, grad_norm_dic, all_log_metrics
 
+    def run_training_teardown(self):
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]: what about making it protected `_teardown_training`?

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (jeremyjordan) [reply pytorch_lightning/trainer/training_loop.py]: nope, sorry that's a weakness of the Github diff renderer. the only change is a tab indent for the `try` block, then code changes at the bottom for teardown.

File: pytorch_lightning/trainer/training_loop.py (hunk 5 above)
Comment (jeremyjordan) [reply pytorch_lightning/trainer/training_loop.py]: what's the benefit? i think it's cleaner to be symmetric with the other methods (eg. `run_training_epoch`, `run_training_batch`, `run_training_teardown`)

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (jeremyjordan) [reply pytorch_lightning/trainer/training_loop.py]: i agree it's a heavy loop, but i kept this as it was on `master`. refactoring the training loop should be a separate discussion+PR

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (jeremyjordan) [reply pytorch_lightning/trainer/trainer.py]: the profiler report is called within `train()`, i felt like that was the more logical place to find the code so i moved it into the `run_training_teardown()` method.

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py]: yes, it was just a heavy for review since all it is marked as change but in fact, there is no change... May I ask to create an issue with this cut-out training and mark is "good first issue"

File: pytorch_lightning/trainer/training_loop.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py]: that make sense... maybe I would start a discussion if these kinds of workaround methods shall start with run_... but maybe its just my weakness in English :D

Comment (jeremyjordan) [reply pytorch_lightning/trainer/training_loop.py]: haha, no worries 😄

**Comment by jeremyjordan** [comment]
@Borda updated to fix merge conflicts with master. were there any more changes requested?

**Comment by Borda** [comment]
@jeremyjordan just fixed another conflict, pls check it...

**Comment by jeremyjordan** [comment]
@Borda looks good! thanks for addressing that

**Comment by williamFalcon** [comment]
@jeremyjordan rebase?

**Comment by jeremyjordan** [comment]
@williamFalcon rebased and ready for review!
